### PR TITLE
Dashboard code with CAN integration.

### PR DESCRIPTION
**Pull request by joshua2020181**

### CODE REVIEW

File: ArduinoCode/Driver-IO/dashboard/dashboard.ino [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+   Turns IMD indicator on or off depending on if IMD_INPUT_PIN detects a voltage.
+*/
+void imdDetection() {
+  if (digitalRead(IMD_INPUT_PIN)) {
+    digitalWrite(IMD_INDICATOR_PIN, HIGH);
+  } else {
+    digitalWrite(IMD_INDICATOR_PIN, LOW);
+  }
+}
+
```
Comment (jamescd18) [inline ArduinoCode/Driver-IO/dashboard/dashboard.ino:81]: These two methods could technically be refactored for an abstraction, whereby they each call a 3rd function and pass an input and an output, but this not required by any means and @mccauleyma will probably say it's very unnecessary. That said, reduction of duplicate code is a principle of good software development. I've provided the example below primarily for your learning.
```
void pinDetection(input, output) {
   if (digitalRead(input)) {
     digitalWrite(output, HIGH);
   } else {
     digitalWrite(output, LOW);
   }
}

void imdDetection() {
   pinDetection(IMD_INPUT_PIN, IMD_INDICATOR_PIN);
}

void bmsDetection() {
   pinDetection(BMS_INPUT_PIN, BMS_INDICATOR_PIN);
}
```

File: ArduinoCode/Driver-IO/dashboard/dashboard.ino [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+const int BMS_INPUT_PIN = 6; // Pin for detecting BMS status
+
+const int IMD_INDICATOR_PIN = 8; // Pin for IMD indicator
+const int BMS_INDICATOR_PIN = 9; // Pin for BMS indicator
+const int BUTTON_LED_PIN = 11; // Pin for integrated LED on the button.
+const int SPEAKER_PIN = 10; // Pin for speaker or relay for speaker (must be pwm pin).
+
+bool shutdownVoltage = false;
+unsigned long speakerCooldown = 0; // Use unsigned long to match millis() function. Will overflow in around 50 days.
+unsigned long buttonCooldown = 0; // Use unsigned long to match millis() function. Will overflow in around 50 days.
```
Comment (jamescd18) [inline ArduinoCode/Driver-IO/dashboard/dashboard.ino:29]: Will these end up overflowing? Why?

File: ArduinoCode/Driver-IO/dashboard/dashboard.ino [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    buttonCooldown = millis() + 500; // Set cooldown to 500ms (prevents triggering button multiple times with one press).
+    if (state) { // Car is turned on.
+      analogWrite(SPEAKER_PIN, 128);
+      speakerCooldown = millis() + 3000;
+      CAN.sendMsgBuf(0x01, 0, 1, ON); // Tell pedal box to turn inverter on
+      if (digitalRead(GEAR_SELECTOR_PIN)) { // Send direction to pedal box when car is turned on.
+        CAN.sendMsgBuf(0x02, 0, 1, FORWARD);
+      } else {
+        CAN.sendMsgBuf(0x02, 0, 1, REVERSE);
+      }
```
Comment (jamescd18) [inline ArduinoCode/Driver-IO/dashboard/dashboard.ino:130]: I spy duplicate code. Same ideas from before apply. Good in principle, not necessary in practice.

**Review by jamescd18 — CHANGES_REQUESTED** [review]
Looks pretty good and thoroughly commented, just a few minor edits. I defer to @mccauleyma for a more detailed review.

File: ArduinoCode/Driver-IO/dashboard/dashboard.ino (hunk 2 above)
Comment (mccauleyma) [reply ArduinoCode/Driver-IO/dashboard/dashboard.ino:29]: They would never overflow. Our battery will die within a few hours, so even a 1 day overflow time would be acceptable. @jamescd18 this code will be entirely restarted on power cycle of the car

**Review by mccauleyma — CHANGES_REQUESTED** [review]
MAJOR ISSUE:
You are still implementing a misinterpretation that was made before and resolved. IMD and BMS do not interact with the Arduino at all. I was gonna comment all the stuff that needs to be removed, but it's quite a few lines and functions scattered about. Please check the KiCad schematic for Driver IO or contact @Cameron-Gordon or I for more info

File: ArduinoCode/Driver-IO/dashboard/dashboard.ino (hunk 1 above)
Comment (mccauleyma) [reply ArduinoCode/Driver-IO/dashboard/dashboard.ino:81]: Yep, unnecessary. More importantly though this code will be deleted anyways: https://github.com/Northeastern-Electric-Racing/NER/pull/71#pullrequestreview-527498466

**Review by mccauleyma — APPROVED** [review]

**Review by rydzynskip — APPROVED** [review]
Looks great to me!!

File: ArduinoCode/Driver-IO/dashboard/dashboard.ino [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    return;
+  }
+  if (digitalRead(GEAR_SELECTOR_PIN)) {
+    CAN.sendMsgBuf(0x02, 0, 1, FORWARD); // tell pedal box to set inverter to forward
+    prevDirection = true;
+  } else {
+    CAN.sendMsgBuf(0x02, 0, 1, REVERSE); // tell pedal box to set inverter to reverse
+    prevDirection = false;
+  }
+
```
Comment (jamescd18) [inline ArduinoCode/Driver-IO/dashboard/dashboard.ino:113]: spare line
```suggestion
```

**Review by jamescd18 — CHANGES_REQUESTED** [review]
super minor thing

**Review by jamescd18 — APPROVED** [review]
Looking much better, and all cleaned up. I love it. Fantastic job!

**Pull request merged**